datashader/glyphs.py

```diff
@@ -127,14 +127,30 @@ def _build_draw_line(append, x_mapper, y_mapper):
     @ngjit
     def draw_line(vt, bounds, x0, y0, x1, y1, i, plot_start, clipped,
                   *aggs_and_cols):
-        """Draw a line using Bresenham's algorithm"""
+        """Draw a line using Bresenham's algorithm
+
+        This method maps points with float coordinates onto a pixel grid. The
+        given bounds have an inclusive lower bound and exclusive upper bound.
+        Any pixels drawn on the x-axis or y-axis upper bound are ignored.
+
+        The vertices of the line segment and the bounds are scaled and
+        transformed onto the pixel grid before use in the calculations.
+        """
         sx, tx, sy, ty = vt
-        # Project to pixel space
+        xmin, xmax, ymin, ymax = bounds
+
+        # Project vertices to pixel space
         x0i = int(x_mapper(x0) * sx + tx)
         y0i = int(y_mapper(y0) * sy + ty)
         x1i = int(x_mapper(x1) * sx + tx)
         y1i = int(y_mapper(y1) * sy + ty)
 
+        # Project bounds to pixel space
+        xmin = int(x_mapper(xmin) * sx + tx)
+        xmax = int(x_mapper(xmax) * sx + tx)
+        ymin = int(y_mapper(ymin) * sy + ty)
+        ymax = int(y_mapper(ymax) * sy + ty)
```

Comment:
Maybe I'm not reading the code correctly but could this be done once outside the aggregation function somehow? Seems like doing it repeatedly for each line vertex will add some overhead.